### PR TITLE
Use suffixes for integer literals

### PR DESCRIPTION
**Pull request by kantai**
This changes the integer literal specification for unsigned integers to a _suffix_ rather than a prefix, addressing #1669.

Note: This PR is against the branch `stage/clarity-v0.2.0`. Because this will be a very disruptive change, I think it makes sense to be packaging this together with any other disruptive language changes (e.g., #1443, #1668, #1344, #1638) we may be making in the short term (following something more like a `git-flow` pattern).

### CODE REVIEW

File: src/vm/docs/mod.rs [hunk 1]
```diff
@@ -302,10 +302,12 @@ const EQUALS_API: SpecialAPI = SpecialAPI {
...
     signature: "(is-eq v1 v2...)",
-    description: "Compares the inputted values, returning `true` if they are all equal. Note that _unlike_ the `(and ...)` function, `(is-eq ...)` will _not_ short-circuit.",
+    description: "Compares the inputted values, returning `true` if they are all equal. Note that 
+_unlike_ the `(and ...)` function, `(is-eq ...)` will _not_ short-circuit. All values supplied to
+is-eq _must_ be the same type.",
     example: "(is-eq 1 1) ;; Returns true
-(is-eq 1 false) ;; Returns false
-(is-eq \"abc\" 234 234) ;; Returns false
+(is-eq true false) ;; Returns false
+(is-eq \"abc\" 234 234) ;; Throws type error
```
Comment (njordhov) [inline src/vm/docs/mod.rs:310]: `(is-eq "abc" 234 234)` is actually an _analysis error_, reported as:

    expecting expression of type 'int', found '(buff 3)'

**Comment by njordhov** [comment]
> Because this will be a very disruptive change, I think it makes sense to be packaging this together with any other disruptive language changes (e.g., #1443, #1668, #1344, #1638) we may be making in the short term

#1344 proposes to **eliminate the need for a `u` affix through type inference.** In contrast, this PR so far only moves the affix to the other end of the integer literal. This would be a good opportunity to improve Clarity by eliminating the integer affix.

**Comment by kantai** [comment]
> #1344 proposes to eliminate the need for a u affix through type inference. In contrast, this PR so far only moves the affix to the other end of the integer literal. This would be a good opportunity to improve Clarity by eliminating the integer affix.

Ultimately, both would be necessary, though there are many cases where an affix would not be necessary, we would not eliminate the ability for developers to specify the exact type for a literal.

**Comment by njordhov** [comment]
> even if integer literal inference is implemented and there are many cases where an affix would not be necessary, we would not eliminate the ability for developers to specify the exact type for a literal.

Having these distinct unsigned integer literals is not an _ability_ but a **burden for developers.** In what cases would they still be required after type inference?

**Comment by kantai** [comment]
Unsigned versus signed integers in Clarity enable developers to guarantee that a runtime error will occur in the event of an underflow at 0. This entirely avoids a large class of errors in current smart contracts.

Without a type specification, statements like

(- 1 2)
(define-public (foo) 1)

And others are ambiguous. It’s true that these are toy examples, that we could perhaps find special cases to address, but it is not _generally_ possible to definitely determine whether a supplied literal should be typed as “underflows at 0” or not without a specifier.

**Comment by jcnelson** [comment]
Agreeing with @kantai.  There is a demonstrable use-case for explicitly-unsigned integers in smart contracts: token balances which must never be negative.

**Comment by njordhov** [comment]
> it is not generally possible to definitely determine whether a supplied literal should be typed as “underflows at 0” or not without a specifier.

Integer literals are not mutable but constants, thus cannot "underflow at zero". 

Rather, the _type_ of the integer literal is primarily relevant for function dispatch. Clarity currently overloads arithmetic functions with different implementations depending on the type of integer. But unsigned integer arithmetic can be dangerous, often straddling on the edge of surprising underflows.

For example, this will underflow if `x` is zero in unsigned integer arithmetic, but works as expected for regular integer arithmetic:

```scheme
(* x (- x 1))
```

Despite undeflowing in unsigned integer arithmetic, the result is a nonnegative integer in regular integer arithmetic. Typically it is undesirable to fail on such underflows in the middle of a calculation. However, arithmetic operations on unsigned integers can be emulated using signed integers.

> Unsigned versus signed integers in Clarity enable developers to guarantee that a runtime error will occur in the event of an underflow at 0. This entirely avoids a large class of errors in current smart contracts.

**The goal of guarding against underflow at zero will be better served by making `uint` a _nonnegative integer type_ instead of the current _unsigned integer type_.** The difference is subtle, but most importantly: The nonnegative integer type is a proper subtype of the `int` type, unlike the unsigned integer type.

As a consequence, the same arithmetic functions will work for both integer types, with no need to overload the functions and dispatch depending on the integer type of the arguments, thereby also making distinct unsigned integer literals unnecessary.

**Comment by njordhov** [comment]
> There is a demonstrable use-case for explicitly-unsigned integers in smart contracts: token balances which must never be negative.

@jcnelson This is a good use case and should be well served by making uint a nonnegative integer subtype instead of the current unsigned integer type. 

Overloading arithmetic functions for unsigned integers introduces a new class of errors. For example, this function to find the difference between two balances will fail with an underflow almost half of the time, even if the result of the calculation always would be a nonnegative integer:

```scheme
(define-read-only (difference (balance1 uint) (balance2 uint))
  (abs (- balance1 balance2)))
```

**Comment by kantai** [comment]
> The goal of guarding against underflow at zero will be better served by making uint a nonnegative integer type instead of the current unsigned integer type.

Interesting, yes, I think I agree with this -- I would support using a subtype for nonnegative integers rather than the current uint/int.

**Comment by kantai** [comment]
Closing this PR for now, since the plan to implement #1716 will supersede this.

**Pull request closed without merging**